provisioner/terraform/install.go

```diff
@@ -0,0 +1,85 @@
+package terraform
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/hc-install/product"
+	"github.com/hashicorp/hc-install/releases"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+)
+
+var (
+	// TerraformVersion is the version of Terraform used internally
+	// when Terraform is not available on the system.
+	TerraformVersion = version.Must(version.NewVersion("1.3.0"))
+
+	minTerraformVersion = version.Must(version.NewVersion("1.1.0"))
+	maxTerraformVersion = version.Must(version.NewVersion("1.3.0"))
+
+	terraformMinorVersionMismatch = xerrors.New("Terraform binary minor version mismatch.")
+)
+
+// Install implements a thread-safe, idempotent Terraform Install
+// operation.
+func Install(ctx context.Context, log slog.Logger, dir string, version *version.Version) (string, error) {
+	err := os.MkdirAll(dir, 0750)
+	if err != nil {
+		return "", err
+	}
+
+	versionFilePath := filepath.Join(dir, "version")
+	lock := flock.New(versionFilePath)
+	ok, err := lock.TryLockContext(ctx, time.Millisecond*100)
+	if err != nil {
+		return "", xerrors.Errorf("acquire lock: %w", err)
+	}
+	if !ok {
+		return "", xerrors.Errorf("could not acquire flock for %v", dir)
```

Comment:
Wouldn't this fail to acquire a lock if another is installing in parallel?

Reply:
Only if the context is cancelled, in which case the install would fail anyways.